Implement batching rules for some view ops

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #42324 vmap: temporarily disable support for (most) random functions
* **#42248 Implement batching rules for some view ops**
* #41943 vmap fallback kernel

Including:
- torch.diagonal
- torch.t
- torch.select
- Tensor.expand_as
- Tensor slicing.

Please let me know in the future if it would be easier to review these
separately (I put five operators into this PR because each
implementation is relatively simple).

Test Plan:
- new tests in `test/test_vmap.py`.
- I would like to have a more structured/automated way of testing but
my previous attempts at making something resulted in something very
complicated.

Differential Revision: [D22846273](https://our.internmc.facebook.com/intern/diff/D22846273)

zou3519 committed:
Implement batching rules for some view ops

Including:
- torch.diagonal
- torch.t
- torch.select
- Tensor.expand_as
- Tensor slicing.

Please let me know in the future if it would be easier to review these
separately (I put five operators into this PR because each
implementation is relatively simple).

Test Plan:
- new tests in `test/test_vmap.py`.
- I would like to have a more structured/automated way of testing but
my previous attempts at making something resulted in something very
complicated.

[ghstack-poisoned]

zou3519 committed:
Update on "Implement batching rules for some view ops"


Including:
- torch.diagonal
- torch.t
- torch.select
- Tensor.expand_as
- Tensor slicing.

Please let me know in the future if it would be easier to review these
separately (I put five operators into this PR because each
implementation is relatively simple).

Test Plan:
- new tests in `test/test_vmap.py`.
- I would like to have a more structured/automated way of testing but
my previous attempts at making something resulted in something very
complicated.

[ghstack-poisoned]

ezyang reviewed (approved):
Looks good, these are simple enough.

Have you considered writing some tests that compare the behavior of your backend fallback with the explicit operator? I don't know how I would wire this up, but that might also be an interesting way to check for correctness.

zou3519 commented:
> Have you considered writing some tests that compare the behavior of your backend fallback with the explicit operator? I don't know how I would wire this up, but that might also be an interesting way to check for correctness.

I have considered that. It would be easier to do if there existed an operator database: given some operator, I'd like to query B0 valid inputs for some shape. Then, I can run the explicit operator on each of those B0 inputs and compare the result to vmap on the stacked B0 tensors. Without an op database I would be writing sample inputs for days. cc @mruberry if any of that helps with the operator database work

zou3519 committed:
Update on "Implement batching rules for some view ops"


Including:
- torch.diagonal
- torch.t
- torch.select
- Tensor.expand_as
- Tensor slicing.

Please let me know in the future if it would be easier to review these
separately (I put five operators into this PR because each
implementation is relatively simple).

Test Plan:
- new tests in `test/test_vmap.py`.
- I would like to have a more structured/automated way of testing but
my previous attempts at making something resulted in something very
complicated.

Differential Revision: [D22846273](https://our.internmc.facebook.com/intern/diff/D22846273)

[ghstack-poisoned]

zou3519 committed:
Update on "Implement batching rules for some view ops"


Including:
- torch.diagonal
- torch.t
- torch.select
- Tensor.expand_as
- Tensor slicing.

Please let me know in the future if it would be easier to review these
separately (I put five operators into this PR because each
implementation is relatively simple).

Test Plan:
- new tests in `test/test_vmap.py`.
- I would like to have a more structured/automated way of testing but
my previous attempts at making something resulted in something very
complicated.

Differential Revision: [D22846273](https://our.internmc.facebook.com/intern/diff/D22846273)

[ghstack-poisoned]

zou3519 committed:
Update on "Implement batching rules for some view ops"


Including:
- torch.diagonal
- torch.t
- torch.select
- Tensor.expand_as
- Tensor slicing.

Please let me know in the future if it would be easier to review these
separately (I put five operators into this PR because each
implementation is relatively simple).

Test Plan:
- new tests in `test/test_vmap.py`.
- I would like to have a more structured/automated way of testing but
my previous attempts at making something resulted in something very
complicated.

Differential Revision: [D22846273](https://our.internmc.facebook.com/intern/diff/D22846273)

[ghstack-poisoned]
